core/src/solve.jl

```diff
@@ -739,13 +814,21 @@ function formulate!(flow_boundary::FlowBoundary, p::Parameters, t::Float64)::Not
     end
 end
 
-function formulate!(pump::Pump, p::Parameters, storage::AbstractVector{Float64})::Nothing
-    (; connectivity, basin, level_boundary) = p
+function formulate!(
+    node::Union{Pump, Weir},
+    p::Parameters,
+    storage::AbstractVector{Float64},
+)::Nothing
+    (; connectivity, basin) = p
     (; graph_flow, flow) = connectivity
-    (; node_id, active, flow_rate, is_pid_controlled) = pump
+    (; node_id, active, flow_rate, is_pid_controlled) = node
     for (id, isactive, rate, pid_controlled) in
         zip(node_id, active, flow_rate, is_pid_controlled)
-        @assert rate >= 0 "Pump flow rate must be positive, found $rate for Pump #$id"
+        if rate < 0
+            error(
+                "$(typeof(node)) flow rate must be non-negative, found $rate for Pump #$id.",
```

Comment:
This error message is generated for both Weir and Pump, but only mentions "Pump".

Reply:
I refactored the flow rate sign test, because:
1. It is better validate this before the simulation than find out during
2. It was very fiddly to test the previous version